rust/agama-cli/src/logs.rs

```diff
@@ -181,64 +230,96 @@ impl LogItem for LogCmd {
     }
 }
 
-// collect existing / requested paths which should already exist turns them into list of log
-// sources
-fn paths_to_log_sources(paths: &[&str], tmp_dir: &TempDir) -> Vec<Box<dyn LogItem>> {
+// Collect existing / requested paths which should already exist in the system.
+// Turns them into list of log sources
+fn paths_to_log_sources(paths: &Vec<String>, tmp_dir: &TempDir) -> Vec<Box<dyn LogItem>> {
     let mut log_sources: Vec<Box<dyn LogItem>> = Vec::new();
 
-    for path in paths {
+    for path in paths.iter() {
         // assumption: path is full path
         if Path::new(path).try_exists().is_ok() {
-            log_sources.push(Box::new(LogPath::new(path, tmp_dir.path())));
+            log_sources.push(Box::new(LogPath::new(path.as_str(), tmp_dir.path())));
         }
     }
 
     log_sources
 }
 
-// some info can be collected via particular commands only, turn it into log sources
-fn cmds_to_log_sources(commands: &[&str], tmp_dir: &TempDir) -> Vec<Box<dyn LogItem>> {
+// Some info can be collected via particular commands only, turn it into log sources
+fn cmds_to_log_sources(
+    commands: &Vec<(String, String)>,
+    tmp_dir: &TempDir,
+) -> Vec<Box<dyn LogItem>> {
     let mut log_sources: Vec<Box<dyn LogItem>> = Vec::new();
 
-    for cmd in commands {
-        log_sources.push(Box::new(LogCmd::new(cmd, tmp_dir.path())));
+    for cmd in commands.iter() {
+        log_sources.push(Box::new(LogCmd::new(
+            cmd.0.as_str(),
+            cmd.1.as_str(),
+            tmp_dir.path(),
+        )));
     }
 
     log_sources
 }
 
-// compress given directory into a tar archive
+// Compress given directory into a tar archive
 fn compress_logs(tmp_dir: &TempDir, result: &String) -> io::Result<()> {
     let compression = DEFAULT_COMPRESSION.0;
+    let tmp_path = tmp_dir
+        .path()
+        .parent()
+        .ok_or(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Malformed path to temporary directory",
+        ))?
+        .as_os_str()
+        .to_str()
+        .ok_or(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Malformed path to temporary directory",
+        ))?;
```

Comment:
Sorry, but I see a lot of repetition here with four errors that seem to be the same (but they are not). Anyway, if you are not interested in making any distinction, you could rewrite this code to:

```rust
    let path = tmp_dir.path();
    let tmp_path = path
        .parent()
        .map(|p| p.as_os_str().to_str())
        .flatten()
        .ok_or(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Malformed path to temporary directory",
        ))?;
    let dir = path
        .file_name()
        .map(|p| p.to_str())
        .flatten()
        .ok_or(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Malformed path to temporary directory",
        ))?;
```

And I still see some repetition... :-)

Reply:
well I can write slightly different message into each error and there won't be any repetition anymore ;-)

Reply:
btw i slightly doubt that the snippet works ... bcs ```as_os_str().to_str()``` returns ```Option<&str>```, so you need to unpack it ... that's why I have another ```ok_or``` in my code.

Reply:
I tried the code before suggesting the change. [parent](https://doc.rust-lang.org/std/path/struct.Path.html#method.parent) already returns an Option, so I call [map](https://doc.rust-lang.org/std/option/enum.Option.html#method.map) to work with the `Some` case. As you said, `as_os_str().to_str()` returns another option, so you get an `Option<Option<&str>>`. And that's why I call [flatten](https://doc.rust-lang.org/std/option/enum.Option.html?search=flatten#method.flatten), which returns a single `Option<&str>` and allows me to call `ok_or`.

Reply:
well to be honest, I see the suggested code a little bit less readable. However I haven't found anything better so far.

Reply:
I mean ... it can be of course perfectly fine ... but i need to invest into it a bit to make it common to me ;-)

Reply:
on the end after enough reading i found something similar, but one word shorter ;-) So, I'll test it and push if it works ;-)

Reply:
Sure, and_then looks like a better option here. Actually, *Rust by example* [confirms It](https://doc.rust-lang.org/rust-by-example/error/option_unwrap/and_then.html).